garage/tf/models/base.py

```diff
@@ -0,0 +1,313 @@
+        self._name_scope = tf.name_scope("default")
+        self._siblings = {}
+
+    @classmethod
+    def _new_child(cls, other, name):
+        """Class method for constructing child model with the given name."""
+        twin = cls(**other.init_spec[1])
+        twin._name = name
+        twin._is_parent = False
+        twin._default_parameters = other._default_parameters
+        twin._is_built = other._is_built
+        twin._variable_scope = other.variable_scope
+        twin._name_scope = tf.name_scope(name)
+        return twin
+
+    def build(self, *inputs, name=None):
+        """
+        Build a child model with the given input(s).
+
+        ***Do not call tf.global_variable_initializers() after building a model
+        as it will reassign random weights to the model.***
+
+        It uses the fixed variable scope for all child models, to ensure
+        parameter sharing. Different child models should have unique name.
+
+        Only the parent model can be built. This ensures the model tree to only
+        have two level: the parent and the children. Similarly, only the parent
+        model can be pickled. Connectivity does not hold in pickling. The model
+        needs to be built again afterwards.
+
+        Args:
+          name(str): Name of the model, which is also the name scope of the
+            model.
+
+        Raises:
+          ValueError when
+            - A child model with the same name exists
+            - A child model is built.
+            - A parent model wit the same name exists.
+
+        Returns:
+          outputs: Output tensor of the model with the given inputs.
+
+        """
+        if not name:
+            if 'default' in self._siblings:
+                raise ValueError("Subgraph {} already exists!".format(
+                    self.name))
+            with self._name_scope:
+                if not self._is_built:
+                    _variable_scope = self.variable_scope
+                else:
+                    _variable_scope = tf.variable_scope(
+                        self._variable_scope._name_or_scope, reuse=True)
+                with _variable_scope:
+                    self._inputs = inputs
+                    outputs = self._build(*inputs)
+                    self._outputs = outputs
+                    self._siblings['default'] = self
+                    if not self._is_built:
+                        variables = self._get_variables().values()
+                        tf.get_default_session().run(
+                            tf.variables_initializer(variables))
+                        if self._default_parameters:
+                            self.parameters = self._default_parameters
+                        self._is_built = True
+        elif name in self._siblings:
+            raise ValueError("Subgraph {} already exists!".format(name))
+        elif not self._is_parent:
+            raise ValueError("Child of a model should not be built!")
```

Comment:
i think you could make this structural rather than an exception by splitting the interns of the model into two classes.

* `Model` class which holds the variable scope
* Zero or more `Network` (or `Graph` or `SubGraph`) classes which use that variable scope.

Then you can have `build` create and add a graphs to the Model class. It is not possible to call `build` on a graph (only a model can do that).

Reply:
Yes this can also avoid circular reference.